scheduler/queue/worker.go

```diff
@@ -0,0 +1,170 @@
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+type worker struct {
+	jobs              <-chan *WorkUnit
+	queue             *ConcurrentRandomQueue[WorkUnit]
+	resolvedResources chan<- *schema.Resource
+
+	logger            zerolog.Logger
+	caser             *caser.Caser
+	invocationID      string
+	deterministicCQID bool
+	metrics           *metrics.Metrics
+}
+
+func (w *worker) work(ctx context.Context, activeWorkSignal *activeWorkSignal) {
+	for j := range w.jobs {
+		activeWorkSignal.Add()
+
+		w.resolveTable(ctx, j.Table, j.Client, j.Parent)
+
+		activeWorkSignal.Done()
+	}
+}
+
+func newWorker(
+	jobs <-chan *WorkUnit,
+	queue *ConcurrentRandomQueue[WorkUnit],
+	resolvedResources chan<- *schema.Resource,
+
+	logger zerolog.Logger,
+	c *caser.Caser,
+	invocationID string,
+	deterministicCQID bool,
+	m *metrics.Metrics,
+) *worker {
+	return &worker{
+		jobs:              jobs,
+		queue:             queue,
+		resolvedResources: resolvedResources,
+		logger:            logger,
+		caser:             c,
+		deterministicCQID: deterministicCQID,
+		invocationID:      invocationID,
+		metrics:           m,
+	}
+}
+
+func (w *worker) resolveTable(ctx context.Context, table *schema.Table, client schema.ClientMeta, parent *schema.Resource) {
+	clientName := client.ID()
+	ctx, span := otel.Tracer(metrics.OtelName).Start(ctx,
+		"sync.table."+table.Name,
+		trace.WithAttributes(
+			attribute.Key("sync.client.id").String(clientName),
+			attribute.Key("sync.invocation.id").String(w.invocationID),
+		),
+	)
+	defer span.End()
+	logger := w.logger.With().Str("table", table.Name).Str("client", clientName).Logger()
+	startTime := time.Now()
+	if parent == nil { // Log only for root tables, otherwise we spam too much.
+		logger.Info().Msg("top level table resolver started")
+	}
+	tableMetrics := w.metrics.TableClient[table.Name][clientName]
+	tableMetrics.OtelStartTime(ctx, startTime)
+
+	res := make(chan any)
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				stack := fmt.Sprintf("%s\n%s", err, string(debug.Stack()))
+				logger.Error().Interface("error", err).Str("stack", stack).Msg("table resolver finished with panic")
+				tableMetrics.OtelPanicsAdd(ctx, 1)
+				atomic.AddUint64(&tableMetrics.Panics, 1)
+			}
+			close(res)
+		}()
+		if err := table.Resolver(ctx, client, parent, res); err != nil {
+			logger.Error().Err(err).Msg("table resolver finished with error")
+			tableMetrics.OtelErrorsAdd(ctx, 1)
+			atomic.AddUint64(&tableMetrics.Errors, 1)
+			return
+		}
+	}()
+
+	for r := range res {
+		w.resolveResource(ctx, table, client, parent, r)
+	}
+
+	endTime := time.Now()
+	duration := endTime.Sub(startTime)
+	tableMetrics.Duration.Store(&duration)
+	tableMetrics.OtelEndTime(ctx, endTime)
+	if parent == nil {
+		logger.Info().Uint64("resources", tableMetrics.Resources).Uint64("errors", tableMetrics.Errors).Dur("duration_ms", duration).Msg("table sync finished")
```

Comment:
OK now I remember what I removed here. We used to log a message here with the metrics of all relations after a parent finished.

Not we don't do that since reaching this code doesn't mean all relations finished syncing, it only means the parent finished pushing all the new work units into the queue.

See https://github.com/cloudquery/plugin-sdk/blob/2977ae3d373042219e514f744d59cf07af5faea3/scheduler/scheduler_dfs.go#L125

I don't think it's a blocker and we can add logs alter on. But still something to consider